docs/design/eth-subscribe.md

```diff
@@ -0,0 +1,366 @@
+{"jsonrpc":"2.0","id": 2, "method": "eth_subscribe", "params": ["newPendingTransactions"]}
+```
+
+#### Response
+
+This response is sent whenever new data is available.
+
+```javascript
+{"id":1,"result":"0xc3b33aa549fb9a60e95d21862596617c","jsonrpc":"2.0"}
+
+{
+    "jsonrpc":"2.0",
+    "method":"eth_subscription",
+    "params":{
+        "subscription": SUBSCRIPTION_ID,
+        "result":"0xd6fdc5cc41a9959e922f30cb772a9aef46f4daea279307bc5f7024edc4ccd7fa"
+    }
+}
+```
+
+## Mock code
+
+### Server side
+
+```javascript
+const express = require('express');
+const path = require('path');
+const { createServer } = require('http');
+const { WebSocketServer } = require('ws');
+
+const app = express();
+app.use(express.static(path.join(__dirname, '/public')));
+
+const server = createServer(app);
+const wss = new WebSocketServer({ server });
+
+wss.on('connection', function (ws) {
+  ws.on('message', function (msg) {
+    const msgString = msg.toString('base64');
+    const request = JSON.parse(Buffer.from(msgString, 'base64').toString('ascii'));
+    const { method, params } = request;
+
+    if (method === 'eth_subscribe') {
+      const event = params[0];
+
+      if (event === 'log') {
+        const filter = params[1];
+        // Cache the `event`, `filter`, `SUBSCRIPTION_ID` and the current block number values
+
+        // This establishes the subscription
+        ws.send(
+          JSON.stringify({
+            id: request.id,
+            result: SUBSCRIPTION_ID,
+            jsonrpc: '2.0',
+          })
+        );
+      }
+
+      // Handle other types of events: newPendingTransactions, newHeads, syncing
+    }
+
+    // The WebSocketProvider emits an event to `eth_chainId` right after establishing a connection
+    else if (method === 'eth_chainId') {
+      ws.send(
+        JSON.stringify({
+          id: request.id,
+          result: {
+            result: CHAIN_ID,
+          },
+          jsonrpc: '2.0',
+        })
+      );
+    }
+
+    // Handle other methods
+  });
+
+  ws.on('error', console.error);
+
+  ws.on('close', function () {
+    // remove subscriptions from cache
+  });
+});
+```
+
+### Client side
+
+```javascript
+var subscription = web3.eth.subscribe(
+  'logs',
+  {
+    address: '0x123456..',
+    topics: ['0x12345...'],
+  },
+  function (error, result) {
+    if (!error) console.log(result);
+  }
+);
+
+// unsubscribes the subscriptionsubscription.unsubscribe(function(error, success){
+if (success) {
+  console.log('Successfully unsubscribed!');
+}
+```
+
+### Mirror node polling:
+
+1. Maintain a list of all live socket connections. This can be done with a `key-value` store where the `key` is a unique identifier (some ws libraries provide such an id, but if needed it can be generated by the relay) of the socket object, and the `value` is the socket object itself.
+2. Maintain a list of all active subscriptions and the corresponding socket objects. It should have the following structure:
```

Comment:
To confirm is this implying `count(live)` >= `count(active)`.
Also for every new subscription are we picking up from the list of remaining live connections. Meaning live connections always remain open and it's more of a queue system to assign and release usage of it?

Reply:
A single socket connection can have 0, 1 or multiple subscriptions. It is created and destroyed when a client connects/disconnects from the ws-server. Subscriptions are created when `eth_subscribe` is called and are destroyed on `eth_unsubscribe` or when the associated connection is broken.